lib/workload/stateless/stacks/filemanager/docs/ARCHITECTURE.md

```diff
@@ -41,6 +41,18 @@ Within the application code, out of order events are removed within the [events]
 By default, filemanager makes no assumption about the ordering of events, and ingests events in the order that they arrive.
 The sequencer value is stored on the `s3_object` table, which allows ordering entries when querying.
 
+### Current vs historical records
+
+Since the filemanager database keeps growing as records are never deleted, the current state of records is stored on a
+`is_current_state` column. This column indicates which records represent real objects in S3, and which records are
+historical data that represent previously deleted objects.
+
+This value is computed when events are ingested, and automatically kept up to date. This is done at ingestion because
+the performance impact of determining this is too great on every API call. This does incur a performance penalty for
+ingestion. However, it should be minimal as only other current records need to be considered. This is because records
+only need to transition from current to historical (and not the other way around). Records which are current represent
+a smaller subset of all records, meaning that only a smaller part of the dataset needs to be queried when ingesting.
```

Comment:
Not sure I completely follow here.... 
Why do other records need to be queried when a new one is ingested?

>from current to historical (and not the other way around)
How would objects restored from the archive bucket be handled?

Also, perhaps for future consideration, if handling the data within one table get's tricky, consider moving records between tables ? E.g. a current_records vs deleted_records table? (I think the Django history functionality works like that)

Reply:
> Why do other records need to be queried when a new one is ingested?

A record starts as current, and then some logic needs to make it historical.

For example, say `Created` event "A" comes in for a given key/bucket. At that point in time, "A" has `is_current_state` set to true on ingestion. Then, another `Created` event "B" comes in for the same key/bucket, which represents overwriting that object. Now, "B" has `is_current_state` set to true given that it's the new current state. Something needs to flip `is_current_state` to false on "A" as it's no longer current. This requires querying the database, which is what [this](https://github.com/umccr/orcabus/pull/615#discussion_r1815754904) query does on ingestion.

This doesn't have to happen on ingestion, it could happen in another process some time afterwards. However, I think the cost of doing it at ingestion isn't too high so I've added it to make the state of the records as accurate as soon as possible.

Does that make sense (I'll add this in the docs for the next PR)?

> Also, perhaps for future consideration, if handling the data within one table get's tricky, consider moving records between tables ?

Yeah, agreed, that's the idea with task 2 on https://github.com/umccr/orcabus/issues/614. I'll do this by partitioning the table, which more-or-less creates two separate tables, but allows queries to treat it like one table. When this is set up in postgres, it happens automatically on a particular partition key. This is why I'm adding the `is_current_state` here, because this allows the table to be partitioned on that column.

Reply:
Ah, yes. I forgot we are dealing with object versions.... 
Querying with bucket + key + version should be quick and efficient... all good!